File: src/engine/economy.ts
import { Decimal } from './bignum';
import { ResourceId } from './resources';
import { ResourceTotals } from './gameState';
import { GameState } from './gameState';
import {
  ALL_TECHNOLOGIES,
  TECH_BY_ID,
  isTechAvailable,
  bulkPurchaseCost,
  maxAffordableQuantity,
} from './technologies';
import { PrestigeMultipliers } from './prestige';

/**
 * Pricing rule of the whole game, and the one invariant every part of this
 * file exists to protect:
 *
 *   **A price never rises for any reason other than your own purchases of
 *   that exact thing.**
 *
 * Concretely, the amount charged for a technology is a pure function of
 * `(technology, how many of that technology you already own)`, scaled down —
 * never up — by the prestige discount. Nothing about the rest of the
 * civilization enters into it. So:
 *
 * - a one-time unlock, multiplier or choice is quoted once and costs exactly
 *   that forever, however large the tree around it grows;
 * - a generator's *first* unit always costs its listed base price, and only
 *   the units you have personally bought of it make the next one dearer.
 *
 * There used to be a civilization-complexity surcharge here that multiplied
 * every price by the number of distinct technologies owned. It paced the
 * middle of the game, but it did so by quietly re-pricing things the player
 * had already been shown — every new frontier made every *other* frontier
 * more expensive, so the reward for expanding was a bigger bill. Pacing now
 * comes entirely from the tier curves in `constants.ts > BALANCE`, which are
 * baked into each technology's listed price up front and never move.
 */

/** The price actually charged for `nominal`, after any prestige discount. Only ever ≤ `nominal`. */
export function effectiveCostAmount(nominal: Decimal, discount: number): Decimal {
  return discount <= 0 ? nominal : nominal.mul(1 - Math.min(0.9, discount));
}

/**
 * Restates a wallet in the "nominal" units the cost curves are written in, so
 * the closed-form affordability maths in `technologies/index.ts` needs no
 * knowledge of the prestige discount: divide what the player holds by exactly
 * the factor that will multiply the price.
 *
 * Exported because the UI has to answer the same question the engine does —
 * running both off this keeps a Buy button from ever offering a purchase
 * `purchaseTechnology` would then refuse.
 */
export function nominalizeWallet(
  resources: ResourceTotals,
  costs: { resource: string }[],
  discount: number,
): Record<string, Decimal> {
  const scaled: Record<string, Decimal> = {};
  const factor = 1 - Math.min(0.9, Math.max(0, discount));
  for (const c of costs) {
    const balance = resources[c.resource as ResourceId] ?? Decimal.ZERO;
    scaled[c.resource] = factor >= 1 ? balance : balance.div(factor);
  }
  return scaled;
}

export interface PurchaseResult {
  state: GameState;
  purchasedQuantity: number;
  success: boolean;
}

/**
 * Buys up to `requestedQuantity` more units of a technology (1, 10, 100, or
 * `Infinity` for "buy max"). Fails atomically — either the affordable
 * quantity (capped at what's requested) is purchased and paid for in one
 * step, or nothing changes.
 */
export function purchaseTechnology(
  state: GameState,
  techId: string,
  requestedQuantity: number,
  prestige: PrestigeMultipliers,
  disabledTechIds: ReadonlySet<string> = new Set(),
): PurchaseResult {
  const tech = TECH_BY_ID[techId];
  if (!tech || disabledTechIds.has(techId)) return { state, purchasedQuantity: 0, success: false };
  if (!isTechAvailable(tech, state.techOwned)) return { state, purchasedQuantity: 0, success: false };

  const owned = state.techOwned[techId] ?? 0;
  const roomLeft = tech.maxOwned === Infinity ? Infinity : tech.maxOwned - owned;
  if (roomLeft <= 0) return { state, purchasedQuantity: 0, success: false };

  const scaledAvailable = nominalizeWallet(state.resources, tech.cost, prestige.techCostDiscount);

  const cappedQuantity = Math.min(requestedQuantity, roomLeft === Infinity ? Number.MAX_SAFE_INTEGER : roomLeft);
  const quantity = maxAffordableQuantity(tech, owned, scaledAvailable, cappedQuantity);
  if (quantity <= 0) return { state, purchasedQuantity: 0, success: false };

  const totalCost = bulkPurchaseCost(tech, owned, quantity);
  const newResources = { ...state.resources };
  for (const c of totalCost) {
    const charged = effectiveCostAmount(c.amount, prestige.techCostDiscount);
    newResources[c.resource as ResourceId] = newResources[c.resource as ResourceId].sub(charged).clampMin(0);
  }

  const newTechOwned = { ...state.techOwned, [techId]: owned + quantity };

  return {
    state: { ...state, resources: newResources, techOwned: newTechOwned },
    purchasedQuantity: quantity,
    success: true,
  };
}

/** Grants a technology for free (owned = max(current, 1)) — used for prestige "starting tech" bonuses. */
export function grantTechnology(state: GameState, techId: string): GameState {
  const owned = state.techOwned[techId] ?? 0;
  if (owned > 0) return state;
  return { ...state, techOwned: { ...state.techOwned, [techId]: 1 } };
}

export function listAvailableTechnologies(state: GameState) {
  return ALL_TECHNOLOGIES.filter((t) => isTechAvailable(t, state.techOwned) || (state.techOwned[t.id] ?? 0) > 0);
}

File: app/src/main/java/com/earthgame/idle/domain/economy/Economy.kt
package com.earthgame.idle.domain.economy

import com.earthgame.idle.domain.engine.GameDecimal
import com.earthgame.idle.domain.model.GameState
import com.earthgame.idle.domain.model.ResourceAmounts
import com.earthgame.idle.domain.model.ResourceId
import com.earthgame.idle.domain.prestige.PrestigeMultipliers
import com.earthgame.idle.domain.technologies.CostLine
import com.earthgame.idle.domain.technologies.TECH_BY_ID
import com.earthgame.idle.domain.technologies.Technology
import com.earthgame.idle.domain.technologies.bulkPurchaseCost
import com.earthgame.idle.domain.technologies.isTechAvailable
import com.earthgame.idle.domain.technologies.maxAffordableQuantity
import kotlin.math.max
import kotlin.math.min

/**
 * Pricing rule of the whole game, and the one invariant this file exists to
 * protect:
 *
 *   **A price never rises for any reason other than your own purchases of that
 *   exact thing.**
 *
 * Concretely, the amount charged for a technology is a pure function of
 * `(technology, how many of that technology you already own)`, scaled down —
 * never up — by the prestige discount. Nothing about the rest of the
 * civilization enters into it. So:
 *
 * - a one-time unlock, multiplier or choice is quoted once and costs exactly
 *   that forever, however large the tree around it grows;
 * - a generator's *first* unit always costs its listed base price, and only the
 *   units you have personally bought of it make the next one dearer.
 *
 * The reference implementation once had a civilization-complexity surcharge
 * that multiplied every price by the number of distinct technologies owned. It
 * paced the middle of the game, but it did so by quietly re-pricing things the
 * player had already been shown — every new frontier made every *other*
 * frontier more expensive, so the reward for expanding was a bigger bill. It is
 * gone, and pacing comes entirely from the tier curves in `Constants.kt`, which
 * are baked into each technology's listed price up front and never move.
 * `EconomyParityTest` locks the invariant in.
 */

/** The price actually charged for [nominal] after any prestige discount. Only ever <= [nominal]. */
fun effectiveCostAmount(nominal: GameDecimal, discount: Double): GameDecimal =
    if (discount <= 0) nominal else nominal * (1.0 - min(0.9, discount))

/**
 * Restates a wallet in the "nominal" units the cost curves are written in, so
 * the closed-form affordability maths needs no knowledge of the prestige
 * discount: divide what the player holds by exactly the factor that will
 * multiply the price.
 *
 * The UI has to answer the same question the engine does, and running both off
 * this keeps a Buy button from ever offering a purchase [purchaseTechnology]
 * would then refuse.
 */
fun nominalizeWallet(
    resources: ResourceAmounts,
    costs: List<com.earthgame.idle.domain.technologies.TechCost>,
    discount: Double,
): Map<ResourceId, GameDecimal> {
    val factor = 1.0 - min(0.9, max(0.0, discount))
    val scaled = mutableMapOf<ResourceId, GameDecimal>()
    for (cost in costs) {
        val balance = resources[cost.resource]
        scaled[cost.resource] = if (factor >= 1.0) balance else balance / factor
    }
    return scaled
}

data class PurchaseResult(
    val state: GameState,
    val purchasedQuantity: Int,
    val success: Boolean,
) {
    companion object {
        fun failed(state: GameState) = PurchaseResult(state, 0, false)
    }
}

/** "Buy max" — capped by what a run can realistically own rather than by an unbounded loop. */
const val BUY_MAX_QUANTITY = Int.MAX_VALUE

/**
 * Buys up to [requestedQuantity] more units of a technology. Fails atomically:
 * either the affordable quantity (capped at what was requested) is purchased
 * and paid for in one step, or nothing changes.
 */
fun purchaseTechnology(
    state: GameState,
    techId: String,
    requestedQuantity: Int,
    prestige: PrestigeMultipliers,
    disabledTechIds: Set<String> = emptySet(),
): PurchaseResult {
    val tech = TECH_BY_ID[techId] ?: return PurchaseResult.failed(state)
    if (techId in disabledTechIds) return PurchaseResult.failed(state)
    if (!isTechAvailable(tech, state.techOwned)) return PurchaseResult.failed(state)

    val owned = state.techOwned[techId] ?: 0
    val roomLeft = tech.maxOwned - owned
    if (roomLeft <= 0) return PurchaseResult.failed(state)

    val scaledAvailable = nominalizeWallet(state.resources, tech.cost, prestige.techCostDiscount)
    val cappedQuantity = min(requestedQuantity, roomLeft)
    val quantity = maxAffordableQuantity(tech, owned, scaledAvailable, cappedQuantity)
    if (quantity <= 0) return PurchaseResult.failed(state)

    val totalCost = bulkPurchaseCost(tech, owned, quantity)
    val resources = state.resources.toBuilder()
    for (line in totalCost) {
        val charged = effectiveCostAmount(line.amount, prestige.techCostDiscount)
        resources[line.resource] = (resources[line.resource] - charged).clampMin(GameDecimal.ZERO)
    }

    return PurchaseResult(
        state = state.copy(
            resources = resources.build(),
            techOwned = state.techOwned + (techId to owned + quantity),
        ),
        purchasedQuantity = quantity,
        success = true,
    )
}

/** Grants a technology for free (owned = max(current, 1)) — the prestige "starting tech" bonus. */
fun grantTechnology(state: GameState, techId: String): GameState {
    val owned = state.techOwned[techId] ?: 0
    if (owned > 0) return state
    return state.copy(techOwned = state.techOwned + (techId to 1))
}

/** The price the player is shown for a given buy quantity, discount included. */
fun quotedCost(tech: Technology, owned: Int, quantity: Int, discount: Double): List<CostLine> =
    bulkPurchaseCost(tech, owned, quantity).map { CostLine(it.resource, effectiveCostAmount(it.amount, discount)) }

/**
 * Seconds until the player can afford [cost] at their current income.
 *
 * A price being saved toward is far more motivating with a clock on it, and
 * because prices in this game never move the clock is a promise rather than an
 * estimate: wait that long and the thing is yours at exactly the number on the
 * button. Null means nothing is producing the resource at all, which the UI
 * renders as "no income" rather than as an infinite countdown.
 */
fun secondsUntilAffordable(
    cost: List<CostLine>,
    resources: ResourceAmounts,
    perSecond: ResourceAmounts,
): Double? {
    var worst = 0.0
    for (line in cost) {
        val held = resources[line.resource]
        val shortfall = line.amount - held
        if (shortfall.lte(GameDecimal.ZERO)) continue
        val rate = perSecond[line.resource]
        if (rate.lte(GameDecimal.ZERO)) return null
        worst = max(worst, (shortfall / rate).toDouble())
    }
    return worst
}
